<a href="https://colab.research.google.com/github/edaUsha/SpaceShip-Titanic/blob/main/Spaceship_Titanic.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
#import warnings
#warnings.filterwarnings('ignore')

In [ ]:
np.random.seed(0)

In [ ]:
test=pd.read_csv('test (1).csv')
train=pd.read_csv('train (1).csv')

In [ ]:
train.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

In [ ]:
test.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0013_01      Earth      True  G/3/S  TRAPPIST-1e  27.0  False   
1     0018_01      Earth     False  F/4/S  TRAPPIST-1e  19.0  False   
2     0019_01     Europa      True  C/0/S  55 Cancri e  31.0  False   
3     0021_01     Europa     False  C/1/S  TRAPPIST-1e  38.0  False   
4     0023_01      Earth     False  F/5/S  TRAPPIST-1e  20.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck              Name  
0          0.0        0.0           0.0     0.0     0.0   Nelly Carsoning  
1          0.0        9.0           0.0  2823.0     0.0    Lerome Peckers  
2          0.0        0.0           0.0     0.0     0.0   Sabih Unhearfus  
3          0.0     6652.0           0.0   181.0   585.0  Meratz Caltilter  
4         10.0        0.0         635.0     0.0     0.0   Brence Harperez

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4277 entries, 0 to 4276
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   4277 non-null   object 
 1   HomePlanet    4190 non-null   object 
 2   CryoSleep     4184 non-null   object 
 3   Cabin         4177 non-null   object 
 4   Destination   4185 non-null   object 
 5   Age           4186 non-null   float64
 6   VIP           4184 non-null   object 
 7   RoomService   4195 non-null   float64
 8   FoodCourt     4171 non-null   float64
 9   ShoppingMall  4179 non-null   float64
 10  Spa           4176 non-null   float64
 11  VRDeck        4197 non-null   float64
 12  Name          4183 non-null   object 
dtypes: float64(6), object(7)
memory usage: 434.5+ KB


In [ ]:
train.isnull().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

In [ ]:
test.isnull().sum()

PassengerId       0
HomePlanet       87
CryoSleep        93
Cabin           100
Destination      92
Age              91
VIP              93
RoomService      82
FoodCourt       106
ShoppingMall     98
Spa             101
VRDeck           80
Name             94
dtype: int64

Checking which statistical measure is better

In [ ]:
train['FoodCourt'].mean()

np.float64(458.07720329024676)

In [ ]:
train['RoomService'].mean()

np.float64(224.687617481203)

In [ ]:
train['ShoppingMall'].mean()

np.float64(173.72916912197996)

In [ ]:
train['Spa'].mean()

np.float64(311.1387779083431)

In [ ]:
train['VRDeck'].mean()

np.float64(304.8547912992357)

mean is the best for the numeric variables

In [ ]:
train_num_cols=['Age','RoomService','ShoppingMall','Spa','VRDeck','FoodCourt']
test_num_cols=['Age','RoomService','ShoppingMall','Spa','VRDeck','FoodCourt']

In [ ]:
train[train_num_cols]=train[train_num_cols].fillna(train[train_num_cols].mean())

In [ ]:
test[test_num_cols]=test[test_num_cols].fillna(test[test_num_cols].mean())

In [ ]:
train.columns

Index(['PassengerId', 'HomePlanet', 'CryoSleep', 'Cabin', 'Destination', 'Age',
       'VIP', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa', 'VRDeck',
       'Name', 'Transported'],
      dtype='object')

In [ ]:
train['Name'].mode()

0     Alraium Disivering
1     Ankalik Nateansive
2            Anton Woody
3              Apix Wala
4          Asch Stradick
5       Carry Contrevins
6            Cuses Pread
7             Dia Cartez
8     Elaney Webstephrey
9      Glena Hahnstonsen
10      Glenna Valezaley
11        Gollux Reedall
12           Grake Porki
13         Gwendy Sykess
14     Juane Popelazquez
15          Keitha Josey
16       Loree Wolfernan
17       Sharie Gallenry
18            Sus Coolez
19      Troya Schwardson
Name: Name, dtype: object

Imputing the categorical columns

In [ ]:
train_cat_cols=['HomePlanet', 'CryoSleep', 'Cabin', 'Destination','VIP']
test_cat_cols=['HomePlanet', 'CryoSleep', 'Cabin', 'Destination','VIP']


In [ ]:
for col in train_cat_cols:
  if train[col].isnull().any():
    train[col]=train[col].fillna(train[col].mode()[0])

/tmp/ipython-input-544-2025272580.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  train[col]=train[col].fillna(train[col].mode()[0])


In [ ]:
for col in test_cat_cols:
  if test[col].isnull().any():
    test[col]=test[col].fillna(test[col].mode()[0])

/tmp/ipython-input-545-617878442.py:3: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  test[col]=test[col].fillna(test[col].mode()[0])


All the columns have been imputed except the Name Column.

In [ ]:
#Name column is also being imputed
train['Name']=train['Name'].fillna("Not Given")
test['Name']=test['Name'].fillna("Not Given")

In [ ]:
print(train.isnull().sum())
print()
print(test.isnull().sum())

PassengerId     0
HomePlanet      0
CryoSleep       0
Cabin           0
Destination     0
Age             0
VIP             0
RoomService     0
FoodCourt       0
ShoppingMall    0
Spa             0
VRDeck          0
Name            0
Transported     0
dtype: int64

PassengerId     0
HomePlanet      0
CryoSleep       0
Cabin           0
Destination     0
Age             0
VIP             0
RoomService     0
FoodCourt       0
ShoppingMall    0
Spa             0
VRDeck          0
Name            0
dtype: int64


All values have been imputed

**EDA**

In [ ]:
spent_cols=['RoomService','ShoppingMall','Spa','VRDeck','FoodCourt']

train['Total Spend']=train[spent_cols].sum(axis=1)
test['Total Spend']=test[spent_cols].sum(axis=1)

In [ ]:
train.head()

PassengerId HomePlanet  CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa      False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth      False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa      False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa      False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth      False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  Total Spend  
0        False          0.0  
1         True        736.0  
2        False      10383.0  
3        False       5176.0  
4         True       1091.0

Discretizing Age in both datas

In [ ]:
from sklearn.preprocessing import KBinsDiscretizer

train_copy=train.copy()
test_copy=test.copy()

kbin=KBinsDiscretizer(n_bins=6,encode='ordinal',strategy='uniform')

train_copy['Age']=kbin.fit_transform(train_copy[['Age']])
test_copy['Age']=kbin.fit_transform(test[['Age']])

In [ ]:
plt.figure(figsize=(5,5))
sns.countplot(y='Age',data=train_copy)
plt.title("Age Discretizationin train")
plt.show()

sns.countplot(y='Age',data=test_copy)
plt.title("Age Discretizationin test")
plt.show()

KeyboardInterrupt: 

In [ ]:
# @title HomePlanet

from matplotlib import pyplot as plt
import seaborn as sns
train.groupby('HomePlanet').size().plot(kind='barh', color=sns.palettes.mpl_palette('Dark2'))
plt.gca().spines[['top', 'right',]].set_visible(False)

# Task
Perform additional feature engineering on the provided data, including splitting the 'cabin' column, creating a 'family size' feature, encoding categorical features, scaling numerical features, and selecting relevant features.

### Subtask:
Extract the deck, number, and side from the 'Cabin' column in both the training and testing datasets.


In [ ]:
train[['Cabin Deck','Cabin Number','Cabin Side']]=train['Cabin'].str.split('/',expand=True)
test[['Cabin Deck','Cabin Number','Cabin Side']]=test['Cabin'].str.split('/',expand=True)


train['Cabin Number']=pd.to_numeric(train['Cabin Number'])
test['Cabin Number']=pd.to_numeric(test['Cabin Number'])

In [ ]:
train['PassengerId'].head()

In [ ]:
# Extract the group number from 'PassengerId'
train['Group'] = train['PassengerId'].apply(lambda x: x.split('_')[0])
test['Group'] = test['PassengerId'].apply(lambda x: x.split('_')[0])

# Count the number of people in each group
train['Family Size'] = train['Group'].map(train['Group'].value_counts())
test['Family Size'] = test['Group'].map(test['Group'].value_counts())

# Display the first few rows with the new feature
display(train.head())
display(test.head())

In [ ]:
a=['Age', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa', 'VRDeck', 'Total Spend','Cabin Number','Family Size','PassengerId']

sns.heatmap(train[a].corr(),annot=True,cmap='coolwarm')

All the graphs

In [ ]:
# prompt: plot all the categorical graphs in test and train

# List of categorical columns to plot
categorical_cols_to_plot = ['HomePlanet', 'CryoSleep', 'Destination', 'VIP', 'Cabin Deck','Family Size']

# Plotting categorical distributions for train data
print("Categorical plots for train data:")
for col in categorical_cols_to_plot:
    plt.figure(figsize=(4, 5))
    sns.countplot(y=col, data=train, palette='viridis', hue=col, legend=False) # Changed palette here
    plt.title(f'Distribution of {col} in train data')
    plt.show()

# Plotting categorical distributions for test data
print("\nCategorical plots for test data:")
for col in categorical_cols_to_plot:
    plt.figure(figsize=(8, 5))
    sns.countplot(y=col, data=test, palette='viridis', hue=col, legend=False) # Changed palette here
    plt.title(f'Distribution of {col} in test data')
    plt.show()

OneHot Encoding and Label Encoding

In [ ]:
train.head()

In [ ]:
train['Group']=pd.to_numeric(train['Group'])

In [ ]:
#train.drop(['Cabin','PassengerId'],axis=1)
#make sure to drop the columns while modelling

In [ ]:
label_cols=['HomePlanet','Destination','Cabin Deck','Cabin Side']
onehot_cols=['CryoSleep','VIP']

In [ ]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder

le=LabelEncoder()
ohe=OneHotEncoder(sparse_output=False,drop='first')

In [ ]:
#label encoding
for col in label_cols:
  train[col]=le.fit_transform(train[col])
  test[col]=le.transform(test[col])

In [ ]:
#one hot encoding
for col in onehot_cols:
  encoded_train=ohe.fit_transform(train[[col]])
  encoded_train_df=pd.DataFrame(encoded_train,columns=ohe.get_feature_names_out([col]))
  train=pd.concat([train,encoded_train_df],axis=1)

  encoded_test=ohe.transform(test[[col]])
  encoded_test_df=pd.DataFrame(encoded_test,columns=ohe.get_feature_names_out([col]))
  test=pd.concat([test,encoded_test_df],axis=1)


In [ ]:
train_transport=ohe.fit_transform(train[['Transported']])
train_transport_df=pd.DataFrame(train_transport,columns=ohe.get_feature_names_out(['Transported']))
train=pd.concat([train,train_transport_df],axis=1)


In [ ]:
#dropping the onehot columns
train.drop(onehot_cols,axis=1,inplace=True)

In [ ]:
test.drop(onehot_cols,axis=1,inplace=True)

Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report,confusion_matrix
from sklearn.preprocessing import StandardScaler

In [ ]:
# Define features (X) and target (y) for training data
X = train.drop(['PassengerId', 'Cabin', 'Name', 'Transported','Transported_True'], axis=1)
y = train['Transported_True']

# Define features for test data, ensuring the same columns as X
x_test = test.drop(['PassengerId', 'Cabin', 'Name', 'Group'], axis=1)

# Initialize the StandardScaler
#scale = StandardScaler()

# Fit the scaler on the training features and transform both training and test features
#X_scale = scale.fit_transform(X)
#x_scale = scale.transform(x_test)

In [ ]:

# Split the scaled training data for model training and evaluation
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3, random_state=42)

# Initialize and train the Random Forest Classifier
model = RandomForestClassifier(n_estimators=1000,criterion='gini', max_depth=15,random_state=42)
model.fit(X_train, y_train)

# Make predictions on the scaled Validation set
y_pred = model.predict(X_val)

# Evaluate the model
accuracy = accuracy_score(y_val, y_pred)
conf_matrix = confusion_matrix(y_val, y_pred)
class_report = classification_report(y_val, y_pred)

print(f"Accuracy: {accuracy}")
print("\nConfusion Matrix:")
print(conf_matrix)
print("\nClassification Report:")
print(class_report)

In [ ]:
#with max depth=0.7929447852760736
#without maxdepth=Accuracy: 0.7940950920245399

0.7944785276073619

In [ ]:
import lightgbm as lgb

# Initialize and train the LightGBM Classifier
lgb_model = lgb.LGBMClassifier(random_state=42)
lgb_model.fit(X_train, y_train)

# Make predictions on the scaled Validation set
y_pred_lgb = lgb_model.predict(X_val)

# Evaluate the model
accuracy_lgb = accuracy_score(y_val, y_pred_lgb)
conf_matrix_lgb = confusion_matrix(y_val, y_pred_lgb)
class_report_lgb = classification_report(y_val, y_pred_lgb)

print(f"LightGBM Accuracy: {accuracy_lgb}")
print("\nLightGBM Confusion Matrix:")
print(conf_matrix_lgb)
print("\nLightGBM Classification Report:")
print(class_report_lgb)

In [ ]:
print("Current features used in the model:")
print(X.columns)
print("\nAll columns in the training data:")
print(train.columns)

# Task
Tune the hyperparameters of a Random Forest and a LightGBM model using grid search with cross-validation and compare their performance.

## Define the hyperparameter grid

### Subtask:
for both the Random Forest and LightGBM models.


**Reasoning**:
Define the hyperparameter grids for Random Forest and LightGBM models as dictionaries.



In [ ]:
param_grid_rf = {
    'n_estimators': [100, 200, 500],
    'max_depth': [10, 15, 20],
    'criterion': ['gini', 'entropy']
}

param_grid_lgb = {
    'n_estimators': [100, 200, 500],
    'max_depth': [10, 15, 20],
    'learning_rate': [0.01, 0.05, 0.1]
}

## Perform grid search with cross-validation

### Subtask:
Perform grid search with cross-validation on the training data for the Random Forest model using the defined parameter grid.


**Reasoning**:
Perform grid search with cross-validation on the training data for the Random Forest model using the defined parameter grid.



In [ ]:
from sklearn.model_selection import GridSearchCV

rf_model = RandomForestClassifier(random_state=42)

grid_search_rf = GridSearchCV(estimator=rf_model, param_grid=param_grid_rf, cv=5,n_jobs=-1,
                               scoring='accuracy')

grid_search_rf.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42), n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [10, 15, 20],
                         'n_estimators': [100, 200, 500]},
             scoring='accuracy')

### Subtask:
Perform grid search with cross-validation on the training data for the LightGBM model using the defined parameter grid.

**Reasoning**:
Perform grid search with cross-validation on the training data for the LightGBM model using the defined parameter grid.

In [ ]:
lgb_model = lgb.LGBMClassifier(random_state=42)

grid_search_lgb = GridSearchCV(estimator=lgb_model, param_grid=param_grid_lgb, cv=5,n_jobs=-1,
                                 scoring='accuracy')

grid_search_lgb.fit(X_train, y_train)

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 3059, number of negative: 3026
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001054 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 2148
[LightGBM] [Info] Number of data points in the train set: 6085, number of used features: 16
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.502712 -> initscore=0.010846
[LightGBM] [Info] Start training from score 0.010846


GridSearchCV(cv=5, estimator=LGBMClassifier(random_state=42), n_jobs=-1,
             param_grid={'learning_rate': [0.01, 0.05, 0.1],
                         'max_depth': [10, 15, 20],
                         'n_estimators': [100, 200, 500]},
             scoring='accuracy')

# Task
Continue the analysis by displaying the best hyperparameters and scores from the previous grid search, evaluating the tuned models on the validation set, comparing their performance, selecting the best model, making predictions on the test set, and generating a submission file.

## Display best parameters and scores

### Subtask:
Print the best parameters and the best cross-validation score for both the Random Forest and LightGBM models obtained from the grid search.


**Reasoning**:
Print the best hyperparameters and best cross-validation score for both Random Forest and LightGBM models from the grid search results.



In [ ]:
print("Best parameters for Random Forest:", grid_search_rf.best_params_)
print("Best cross-validation score for Random Forest:", grid_search_rf.best_score_)

print("\nBest parameters for LightGBM:", grid_search_lgb.best_params_)
print("Best cross-validation score for LightGBM:", grid_search_lgb.best_score_)

Best parameters for Random Forest: {'criterion': 'entropy', 'max_depth': 15, 'n_estimators': 200}
Best cross-validation score for Random Forest: 0.8046014790468365

Best parameters for LightGBM: {'learning_rate': 0.05, 'max_depth': 20, 'n_estimators': 200}
Best cross-validation score for LightGBM: 0.8065735414954807


## Evaluate tuned models

### Subtask:
Train both models with their best parameters on the training set and evaluate their performance on the validation set using appropriate metrics (e.g., accuracy, confusion matrix, classification report).


**Reasoning**:
Train both models with their best parameters on the training set and evaluate their performance on the validation set using appropriate metrics (e.g., accuracy, confusion matrix, classification report).



In [ ]:
# Train the best Random Forest model and evaluate on the validation set
best_rf_model = grid_search_rf.best_estimator_
best_rf_model.fit(X_train, y_train)
y_pred_rf = best_rf_model.predict(X_val)

print("Random Forest Model Evaluation on Validation Set:")
print(f"Accuracy: {accuracy_score(y_val, y_pred_rf)}")
print("\nConfusion Matrix:")
print(confusion_matrix(y_val, y_pred_rf))
print("\nClassification Report:")
print(classification_report(y_val, y_pred_rf))

# Train the best LightGBM model and evaluate on the validation set
best_lgb_model = grid_search_lgb.best_estimator_
best_lgb_model.fit(X_train, y_train)
y_pred_lgb_tuned = best_lgb_model.predict(X_val)

print("\nLightGBM Model Evaluation on Validation Set:")
print(f"Accuracy: {accuracy_score(y_val, y_pred_lgb_tuned)}")
print("\nConfusion Matrix:")
print(confusion_matrix(y_val, y_pred_lgb_tuned))
print("\nClassification Report:")
print(classification_report(y_val, y_pred_lgb_tuned))

Random Forest Model Evaluation on Validation Set:
Accuracy: 0.7952453987730062

Confusion Matrix:
[[1016  273]
 [ 261 1058]]

Classification Report:
              precision    recall  f1-score   support

         0.0       0.80      0.79      0.79      1289
         1.0       0.79      0.80      0.80      1319

    accuracy                           0.80      2608
   macro avg       0.80      0.80      0.80      2608
weighted avg       0.80      0.80      0.80      2608

[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 3059, number of negative: 3026
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001079 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 2148
[LightGBM] [Info] Number of data points in the train set: 6085, number of used features: 16
[LightGBM] [Info] [bina

## Compare model performance

### Subtask:
Compare the performance of the tuned Random Forest and LightGBM models based on the evaluation metrics.


**Reasoning**:
Summarize the performance comparison based on the evaluation metrics.



In [ ]:
print("Performance Comparison of Tuned Models:")
print("\nRandom Forest Model:")
print(f"Accuracy: {accuracy_score(y_val, y_pred_rf)}")
print("Confusion Matrix:\n", confusion_matrix(y_val, y_pred_rf))
print("Classification Report:\n", classification_report(y_val, y_pred_rf))

print("\nLightGBM Model:")
print(f"Accuracy: {accuracy_score(y_val, y_pred_lgb_tuned)}")
print("Confusion Matrix:\n", confusion_matrix(y_val, y_pred_lgb_tuned))
print("Classification Report:\n", classification_report(y_val, y_pred_lgb_tuned))

print("\nSummary of Performance Differences:")
print(f"LightGBM Model achieved a slightly higher accuracy ({accuracy_score(y_val, y_pred_lgb_tuned):.4f}) compared to the Random Forest Model ({accuracy_score(y_val, y_pred_rf):.4f}) on the validation set.")
print("Reviewing the classification reports will provide more detailed insights into precision, recall, and F1-score for each class to understand where the models differ in their predictions.")

Performance Comparison of Tuned Models:

Random Forest Model:
Accuracy: 0.7952453987730062
Confusion Matrix:
 [[1016  273]
 [ 261 1058]]
Classification Report:
               precision    recall  f1-score   support

         0.0       0.80      0.79      0.79      1289
         1.0       0.79      0.80      0.80      1319

    accuracy                           0.80      2608
   macro avg       0.80      0.80      0.80      2608
weighted avg       0.80      0.80      0.80      2608


LightGBM Model:
Accuracy: 0.8052147239263804
Confusion Matrix:
 [[1017  272]
 [ 236 1083]]
Classification Report:
               precision    recall  f1-score   support

         0.0       0.81      0.79      0.80      1289
         1.0       0.80      0.82      0.81      1319

    accuracy                           0.81      2608
   macro avg       0.81      0.81      0.81      2608
weighted avg       0.81      0.81      0.81      2608


Summary of Performance Differences:
LightGBM Model achieved a slight

## Select best model

### Subtask:
Choose the model with the best performance on the validation set as the final model.


**Reasoning**:
Compare the accuracy of the tuned Random Forest and LightGBM models on the validation set and select the one with the higher accuracy as the final model.



In [ ]:
accuracy_rf = accuracy_score(y_val, y_pred_rf)
accuracy_lgb_tuned = accuracy_score(y_val, y_pred_lgb_tuned)

if accuracy_lgb_tuned > accuracy_rf:
  final_model = best_lgb_model
  print(f"LightGBM model selected as the final model with accuracy: {accuracy_lgb_tuned:.4f}")
else:
  final_model = best_rf_model
  print(f"Random Forest model selected as the final model with accuracy: {accuracy_rf:.4f}")

LightGBM model selected as the final model with accuracy: 0.8052


## Make predictions on test data

### Subtask:
Use the selected best model to make predictions on the unseen test dataset.


**Reasoning**:
Use the selected best model to make predictions on the unseen test dataset.



In [ ]:
test_predictions = final_model.predict(test)

ValueError: Number of features of the model must match the input. Model n_features_ is 16 and input n_features is 14

**Reasoning**:
The error indicates a mismatch in the number of features between the training data used to train the model and the test data used for prediction. I need to inspect the columns of both the training features (X) and the test features (x_test) to identify the missing column in x_test and either drop it from X or impute it in x_test.



In [ ]:
print("Columns in training features (X):")
print(X.columns)
print("\nColumns in test features (x_test):")
print(x_test.columns)

Columns in training features (X):
Index(['HomePlanet', 'Destination', 'Age', 'RoomService', 'FoodCourt',
       'ShoppingMall', 'Spa', 'VRDeck', 'Total Spend', 'Cabin Deck',
       'Cabin Number', 'Cabin Side', 'Group', 'Family Size', 'CryoSleep_True',
       'VIP_True'],
      dtype='object')

Columns in test features (x_test):
Index(['HomePlanet', 'Destination', 'Age', 'RoomService', 'FoodCourt',
       'ShoppingMall', 'Spa', 'VRDeck', 'Total Spend', 'Cabin Deck',
       'Cabin Number', 'Cabin Side', 'Family Size', 'CryoSleep_True',
       'VIP_True'],
      dtype='object')


**Reasoning**:
The output confirms that the 'Group' column is present in the training features (X) but missing in the test features (x_test). Since 'Group' was used to derive 'Family Size', and dropping 'Group' from X would require retraining the model, the most straightforward approach is to drop the 'Group' column from X to match the features in x_test.



In [ ]:
X = X.drop('Group', axis=1)

# Re-split the training data with the updated X
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.3, random_state=42)

# Retrain the best LightGBM model with the corrected training data
final_model.fit(X_train, y_train)

# Now, attempt to make predictions on the test set again
test_predictions = final_model.predict(x_test)


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 3059, number of negative: 3026
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001003 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1893
[LightGBM] [Info] Number of data points in the train set: 6085, number of used features: 15
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.502712 -> initscore=0.010846
[LightGBM] [Info] Start training from score 0.010846


## Generate submission file

### Subtask:
Create a submission file in the specified format (e.g., CSV) containing the passenger IDs and their predicted transportation status.


**Reasoning**:
Create a submission file with 'PassengerId' and 'Transported' columns, populate it with test data PassengerIds and predictions, convert 'Transported' to boolean, and save it as a CSV.



In [ ]:
submission_df = pd.DataFrame({'PassengerId': test['PassengerId'], 'Transported': test_predictions})
submission_df['Transported'] = submission_df['Transported'].astype(bool)
submission_df.to_csv('submission.csv', index=False)
display(submission_df.head())

PassengerId  Transported
0     0013_01         True
1     0018_01        False
2     0019_01         True
3     0021_01         True
4     0023_01         True

## Summary:

### Data Analysis Key Findings

*   The best parameters for the Random Forest model were `{'criterion': 'entropy', 'max_depth': 15, 'n_estimators': 200}`, resulting in a cross-validation score of approximately 0.8046.
*   The best parameters for the LightGBM model were `{'learning_rate': 0.05, 'max_depth': 20, 'n_estimators': 200}`, resulting in a cross-validation score of approximately 0.8066.
*   On the validation set, the tuned LightGBM model achieved an accuracy of 0.8052, slightly higher than the tuned Random Forest model's accuracy of 0.7952.
*   Comparing the classification reports on the validation set, the LightGBM model showed slightly better recall and F1-score for the positive class (class 1).
*   The LightGBM model was selected as the final model due to its superior performance on the validation set.
*   A feature mismatch between the training data (containing 'Group') and the test data (lacking 'Group') was identified and resolved by dropping the 'Group' column from the training data before retraining the final model.
*   Predictions were successfully generated for the test dataset using the retrained LightGBM model.
*   A submission file named 'submission.csv' was created with 'PassengerId' and 'Transported' columns, formatted correctly for submission.

### Insights or Next Steps

*   The LightGBM model demonstrated slightly better generalization performance on the validation set compared to Random Forest, likely due to its efficiency and handling of the dataset's characteristics.
*   Further exploration of feature engineering or testing other advanced models could potentially improve the prediction accuracy on the test set.
